File: native/android/app/src/main/java/com/circuvent/app/core/ApiClient.kt
package com.circuvent.app.core

import android.content.Context
import android.content.SharedPreferences
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.util.concurrent.TimeUnit

/**
 * Where the session lives.
 *
 * EncryptedSharedPreferences rather than plain ones: this holds a bearer token
 * that is as good as the password until it expires, on a device that may be
 * rooted or shared. The Expo app gets this for free from its secure store; on
 * this side it is a deliberate choice that has to be made and stated.
 */
class Session(context: Context) {

    /*
     * Declared before `prefs` because the fallback below assigns it while
     * `prefs` is still initialising. Kotlin initialises properties in
     * declaration order, so the other way round compiles as "cannot be
     * initialized before declaration" — which is the compiler catching a real
     * ordering bug rather than being fussy.
     */
    var secure: Boolean = true
        private set

    private val prefs: SharedPreferences = try {
        val key = MasterKey.Builder(context)
            .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
            .build()
        EncryptedSharedPreferences.create(
            context,
            "cv-session",
            key,
            EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
            EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM,
        )
    } catch (e: Exception) {
        /*
         * Keystore failures are real on some OEM builds, and the alternative to
         * a fallback is an app that cannot sign in at all on those handsets.
         * It is a downgrade, so it is recorded rather than hidden: `secure`
         * tells the UI which one it got.
         */
        secure = false
        context.getSharedPreferences("cv-session-plain", Context.MODE_PRIVATE)
    }

    var token: String?
        get() = prefs.getString("token", null)
        set(v) = prefs.edit().apply { if (v == null) remove("token") else putString("token", v) }.apply()

    var refreshToken: String?
        get() = prefs.getString("refresh", null)
        set(v) = prefs.edit().apply { if (v == null) remove("refresh") else putString("refresh", v) }.apply()

    val signedIn: Boolean get() = !token.isNullOrBlank()

    fun clear() = prefs.edit().clear().apply()
}

/** What a call produced: the body, or why there is none. */
sealed class Result<out T> {
    data class Ok<T>(val value: T) : Result<T>()
    data class Err(val message: String, val status: Int = 0) : Result<Nothing>()
}

class ApiClient(private val session: Session) {

    private val http = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .build()

    private val json = Json {
        ignoreUnknownKeys = true   // firmware adds fields; a phone must not break on them
        explicitNulls = false
    }

    /*
     * One refresh at a time.
     *
     * The device list and the home screen fire together, so an expired token
     * produces a burst of 401s. Refresh tokens are single-use, so letting each
     * retry rotate independently means all but one present a spent token — and
     * the server reads a spent token as replay and tears the whole family down,
     * signing the user out of a session that was merely stale. The Expo client
     * learned this the hard way; the lock is here for the same reason.
     */
    private val refreshLock = Mutex()

    suspend fun login(email: String, password: String): Result<AuthResponse> {
        val body = json.encodeToString(
            JsonObject.serializer(),
            buildJsonObject {
                put("email", email)
                put("password", password)
            }
        )
        return when (val r = post(Api.LOGIN, body, authed = false)) {
            is Result.Err -> r
            is Result.Ok -> {
                val auth = runCatching { json.decodeFromString<AuthResponse>(r.value) }.getOrNull()
                    ?: return Result.Err("The server sent something this app could not read.")
                if (auth.token.isBlank()) {
                    Result.Err(auth.error ?: "Those details were not accepted.")
                } else {
                    session.token = auth.token
                    auth.refreshToken?.let { session.refreshToken = it }
                    Result.Ok(auth)
                }
            }
        }
    }

    suspend fun devices(): Result<List<Device>> =
        when (val r = get(Api.DEVICES)) {
            is Result.Err -> r
            is Result.Ok -> runCatching { json.decodeFromString<DeviceList>(r.value).devices }
                .fold({ Result.Ok(it) }, { Result.Err("Could not read the device list.") })
        }

    suspend fun command(deviceId: String, cmd: JsonObject): Result<Unit> =
        when (val r = post(Api.command(deviceId), json.encodeToString(JsonObject.serializer(), cmd))) {
            is Result.Err -> r
            is Result.Ok -> Result.Ok(Unit)
        }

    // ---------------------------------------------------------------- plumbing

    private suspend fun get(path: String) = call(Request.Builder().url(Api.BASE + path).get(), true)

    private suspend fun post(path: String, body: String, authed: Boolean = true) = call(
        Request.Builder()
            .url(Api.BASE + path)
            .post(body.toRequestBody("application/json".toMediaType())),
        authed,
    )

    private suspend fun call(
        builder: Request.Builder,
        authed: Boolean,
        allowRetry: Boolean = true,
    ): Result<String> = withContext(Dispatchers.IO) {
        if (authed) session.token?.let { builder.header("Authorization", "Bearer $it") }
        val req = builder.header("Accept", "application/json").build()

        val res = try {
            http.newCall(req).execute()
        } catch (e: Exception) {
            return@withContext Result.Err("Could not reach Circuvent. Check the connection.")
        }

        res.use {
            val text = it.body?.string().orEmpty()

            if (it.code == 401 && authed && allowRetry && !session.refreshToken.isNullOrBlank()) {
                if (refresh()) {
                    // The builder carries a stale Authorization header, so the
                    // request is rebuilt rather than replayed.
                    return@withContext call(
                        builder.removeHeader("Authorization"), authed, allowRetry = false
                    )
                }
            }

            if (!it.isSuccessful) {
                return@withContext Result.Err(readError(text, it.code), it.code)
            }
            Result.Ok(text)
        }
    }

    private suspend fun refresh(): Boolean = refreshLock.withLock {
        val rt = session.refreshToken ?: return false
        val body = json.encodeToString(
            JsonObject.serializer(),
            buildJsonObject { put("refreshToken", rt) }
        )
        val res = withContext(Dispatchers.IO) {
            runCatching {
                http.newCall(
                    Request.Builder()
                        .url(Api.BASE + Api.REFRESH)
                        .post(body.toRequestBody("application/json".toMediaType()))
                        .build()
                ).execute()
            }.getOrNull()
        } ?: return false

        res.use {
            if (!it.isSuccessful) {
                // A refused refresh means the family is gone. Holding a dead
                // token would leave every later call retrying forever.
                session.clear()
                return false
            }
            val auth = runCatching {
                json.decodeFromString<AuthResponse>(it.body?.string().orEmpty())
            }.getOrNull() ?: return false
            if (auth.token.isBlank()) return false
            session.token = auth.token
            auth.refreshToken?.let { t -> session.refreshToken = t }
            return true
        }
    }

    /** The server's own words when it sends them, never a paraphrase. */
    private fun readError(text: String, code: Int): String {
        val fromBody = runCatching {
            (json.parseToJsonElement(text) as? JsonObject)
                ?.get("error")
                ?.toString()
                ?.trim('"')
        }.getOrNull()
        if (!fromBody.isNullOrBlank() && fromBody != "null") return fromBody
        return when (code) {
            403 -> "This account is not allowed to do that."
            404 -> "That is no longer there."
            in 500..599 -> "Circuvent is having trouble. Try again shortly."
            else -> "That did not work (HTTP $code)."
        }
    }
}

File: native/ios/Circuvent/Core/ApiClient.swift
import Foundation

/// Where the session lives.
///
/// The Keychain rather than UserDefaults: this holds a bearer token that is as
/// good as the password until it expires. UserDefaults is a plist in the app
/// container, readable from a backup, and it is where this sort of thing ends
/// up when nobody makes the decision deliberately.
final class Session {
    private let service = "com.circuvent.app.nativeclient"

    var token: String? {
        get { read("token") }
        set { write("token", newValue) }
    }

    var refreshToken: String? {
        get { read("refresh") }
        set { write("refresh", newValue) }
    }

    var signedIn: Bool { !(token ?? "").isEmpty }

    func clear() {
        write("token", nil)
        write("refresh", nil)
    }

    private func query(_ key: String) -> [String: Any] {
        [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: key,
        ]
    }

    private func read(_ key: String) -> String? {
        var q = query(key)
        q[kSecReturnData as String] = true
        q[kSecMatchLimit as String] = kSecMatchLimitOne
        var out: CFTypeRef?
        guard SecItemCopyMatching(q as CFDictionary, &out) == errSecSuccess,
              let data = out as? Data else { return nil }
        return String(data: data, encoding: .utf8)
    }

    private func write(_ key: String, _ value: String?) {
        let q = query(key)
        SecItemDelete(q as CFDictionary)
        guard let value, let data = value.data(using: .utf8) else { return }
        var add = q
        add[kSecValueData as String] = data
        /*
         * Not synchronised to iCloud, and unavailable until the device has been
         * unlocked once after boot. A session token that follows somebody onto
         * a second device is a session they never started there.
         */
        add[kSecAttrAccessible as String] = kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly
        SecItemAdd(add as CFDictionary, nil)
    }
}

enum ApiError: Error, LocalizedError {
    case unreachable
    case server(String, Int)

    var errorDescription: String? {
        switch self {
        case .unreachable:
            return "Could not reach Circuvent. Check the connection."
        case .server(let message, _):
            return message
        }
    }
}

/// The REST client.
///
/// Deliberately the same shape as the Kotlin `ApiClient`: same endpoints, same
/// single-flight refresh, same rule about never paraphrasing the server's own
/// error text. Two clients that solve the same problem differently are two
/// clients that behave differently under failure, and failure is the only time
/// anybody notices.
actor ApiClient {
    private let session: Session
    private let urlSession: URLSession
    private let decoder: JSONDecoder = {
        let d = JSONDecoder()
        return d
    }()

    /// Set while a refresh is in flight, so a burst of 401s rotates once.
    ///
    /// Refresh tokens are single-use, so letting each retry rotate
    /// independently means all but one present a spent token — and the server
    /// reads a spent token as replay and tears the whole family down, signing
    /// the user out of a session that was merely stale. The Expo client learned
    /// this the hard way.
    private var refreshTask: Task<Bool, Never>?

    init(session: Session) {
        self.session = session
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 30
        config.waitsForConnectivity = true
        self.urlSession = URLSession(configuration: config)
    }

    func login(email: String, password: String) async throws -> AuthResponse {
        let body: [String: JSONValue] = [
            "email": .string(email),
            "password": .string(password),
        ]
        let data = try await send(Api.login, method: "POST", body: body, authed: false)
        let auth = try decoder.decode(AuthResponse.self, from: data)
        guard !auth.token.isEmpty else {
            throw ApiError.server(auth.error ?? "Those details were not accepted.", 401)
        }
        session.token = auth.token
        if let r = auth.refreshToken { session.refreshToken = r }
        return auth
    }

    func devices() async throws -> [Device] {
        let data = try await send(Api.devices, method: "GET")
        return try decoder.decode(DeviceList.self, from: data).devices
    }

    func command(deviceID: String, _ cmd: [String: JSONValue]) async throws {
        _ = try await send(Api.command(deviceID), method: "POST", body: cmd)
    }

    // MARK: - plumbing

    private func send(
        _ path: String,
        method: String,
        body: [String: JSONValue]? = nil,
        authed: Bool = true,
        allowRetry: Bool = true
    ) async throws -> Data {
        guard let url = URL(string: Api.base + path) else { throw ApiError.unreachable }
        var req = URLRequest(url: url)
        req.httpMethod = method
        req.setValue("application/json", forHTTPHeaderField: "Accept")
        if let body {
            req.setValue("application/json", forHTTPHeaderField: "Content-Type")
            req.httpBody = try JSONEncoder().encode(body)
        }
        if authed, let t = session.token {
            req.setValue("Bearer \(t)", forHTTPHeaderField: "Authorization")
        }

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await urlSession.data(for: req)
        } catch {
            throw ApiError.unreachable
        }

        guard let http = response as? HTTPURLResponse else { throw ApiError.unreachable }

        if http.statusCode == 401, authed, allowRetry, session.refreshToken != nil {
            if await refreshOnce() {
                return try await send(path, method: method, body: body,
                                      authed: authed, allowRetry: false)
            }
        }

        guard (200..<300).contains(http.statusCode) else {
            throw ApiError.server(message(from: data, status: http.statusCode), http.statusCode)
        }
        return data
    }

    private func refreshOnce() async -> Bool {
        if let existing = refreshTask { return await existing.value }
        let task = Task<Bool, Never> { [session] in
            guard let rt = session.refreshToken,
                  let url = URL(string: Api.base + Api.refresh) else { return false }
            var req = URLRequest(url: url)
            req.httpMethod = "POST"
            req.setValue("application/json", forHTTPHeaderField: "Content-Type")
            req.httpBody = try? JSONEncoder().encode(["refreshToken": JSONValue.string(rt)])

            guard let (data, response) = try? await URLSession.shared.data(for: req),
                  let http = response as? HTTPURLResponse else { return false }

            guard (200..<300).contains(http.statusCode) else {
                // A refused refresh means the family is gone. Holding a dead
                // token would leave every later call retrying forever.
                session.clear()
                return false
            }
            guard let auth = try? JSONDecoder().decode(AuthResponse.self, from: data),
                  !auth.token.isEmpty else { return false }
            session.token = auth.token
            if let r = auth.refreshToken { session.refreshToken = r }
            return true
        }
        refreshTask = task
        let ok = await task.value
        refreshTask = nil
        return ok
    }

    /// The server's own words when it sends them, never a paraphrase.
    private func message(from data: Data, status: Int) -> String {
        if let obj = try? JSONDecoder().decode([String: JSONValue].self, from: data),
           let text = obj["error"]?.stringValue, !text.isEmpty {
            return text
        }
        switch status {
        case 403: return "This account is not allowed to do that."
        case 404: return "That is no longer there."
        case 500...599: return "Circuvent is having trouble. Try again shortly."
        default: return "That did not work (HTTP \(status))."
        }
    }
}
